File: templates/ruby/runner.rb
require_relative "./config"
require_relative "./spot"

# =============================================================================
# METAFIDE BOT RUNTIME — runner.rb
# =============================================================================
#
# This file is the actual runtime launcher for the bot.
#
# Responsibilities:
#   1. Validate configuration before startup
#   2. Register shutdown signal handlers
#   3. Start the continuous bot loop from spot.rb
#
# This file is used in two modes:
#
#   Foreground mode:
#     ruby index.rb run
#
#   Background mode:
#     ruby index.rb start
#
# Why this file exists:
#   It separates "CLI process management" from "bot runtime logic".
#   That keeps index.rb focused on commands, while this file focuses on
#   actually starting and running the bot safely.
# =============================================================================

# Validates all config values before the bot starts.
#
# Why this matters:
#   If config is invalid, we want to fail immediately with clear messages
#   instead of letting the bot run in a broken or unsafe state.
def validate_config
  errors = []

  # ---------------------------------------------------------------------------
  # Required fields
  # ---------------------------------------------------------------------------
  errors << "METAFIDE_API_KEY is not set" if METAFIDE_API_KEY.to_s.strip.empty?
  errors << "METAFIDE_USER_ADDRESS is not set" if METAFIDE_USER_ADDRESS.to_s.strip.empty?

  # ---------------------------------------------------------------------------
  # NETWORK validation
  # ---------------------------------------------------------------------------
  valid_networks = %w[mainnet testnet]
  unless valid_networks.include?(NETWORK)
    errors << %(NETWORK must be one of: #{valid_networks.join(", ")} — got "#{NETWORK}")
  end

  # ---------------------------------------------------------------------------
  # INTERVAL validation
  # ---------------------------------------------------------------------------
  valid_intervals = [60, 3600, 23400, 86400]
  unless valid_intervals.include?(INTERVAL)
    errors << %(INTERVAL must be one of: #{valid_intervals.join(", ")} — got "#{INTERVAL}")
  end

  # ---------------------------------------------------------------------------
  # Currency and asset validation
  # ---------------------------------------------------------------------------
  errors << %(CURRENCY must be "USDC" — got "#{CURRENCY}") unless CURRENCY == "USDC"
  errors << %(ASSET must be "BTC_USDT" — got "#{ASSET}") unless ASSET == "BTC_USDT"

  # ---------------------------------------------------------------------------
  # Position and feature flag validation
  # ---------------------------------------------------------------------------
  unless MAX_ALLOWED_POSITIONS.is_a?(Numeric) &&
         MAX_ALLOWED_POSITIONS >= 1 &&
         MAX_ALLOWED_POSITIONS <= 10
    errors << %(MAX_ALLOWED_POSITIONS must be a number between 1 and 10 — got "#{MAX_ALLOWED_POSITIONS}")
  end

  unless ENABLE_EARLY_PRECISION == true || ENABLE_EARLY_PRECISION == false
    errors << %(ENABLE_EARLY_PRECISION must be true or false — got "#{ENABLE_EARLY_PRECISION}")
  end

  # ---------------------------------------------------------------------------
  # POSITIONS_RANGE validation
  # Each interval must have an allowed set of position sizes.
  # ---------------------------------------------------------------------------
  position_minimums = {
    60 => 0.1,
    3600 => 1,
    23400 => 5,
    86400 => 5
  }

  valid_intervals.each do |interval|
    range_values = POSITIONS_RANGE[interval]

    if !range_values.is_a?(Array) || range_values.empty?
      errors << "POSITIONS_RANGE[#{interval}] must be a non-empty array"
    else
      minimum = position_minimums[interval]
      invalid = range_values.select { |value| !value.is_a?(Numeric) || value < minimum }

      if invalid.any?
        errors << "POSITIONS_RANGE[#{interval}] values must be numbers >= #{minimum} — invalid: [#{invalid.join(", ")}]"
      end
    end
  end

  extra_position_keys = POSITIONS_RANGE.keys - valid_intervals
  if extra_position_keys.any?
    errors << "POSITIONS_RANGE has unexpected keys: [#{extra_position_keys.join(", ")}] — only #{valid_intervals.join(", ")} are allowed"
  end

  # ---------------------------------------------------------------------------
  # PRICE_RANGES validation
  # Each interval must define a numeric min/max randomization range.
  # ---------------------------------------------------------------------------
  valid_intervals.each do |interval|
    price_range = PRICE_RANGES[interval]

    if !price_range.is_a?(Hash)
      errors << "PRICE_RANGES[#{interval}] must be an object with { min, max }"
    else
      errors << "PRICE_RANGES[#{interval}].min must be a number" unless price_range[:min].is_a?(Numeric)
      errors << "PRICE_RANGES[#{interval}].max must be a number" unless price_range[:max].is_a?(Numeric)

      if price_range[:min].is_a?(Numeric) &&
         price_range[:max].is_a?(Numeric) &&
         price_range[:min] >= price_range[:max]
        errors << "PRICE_RANGES[#{interval}].min must be less than .max"
      end
    end
  end

  extra_price_keys = PRICE_RANGES.keys - valid_intervals
  if extra_price_keys.any?
    errors << "PRICE_RANGES has unexpected keys: [#{extra_price_keys.join(", ")}] — only #{valid_intervals.join(", ")} are allowed"
  end

  # ---------------------------------------------------------------------------
  # Final result
  # ---------------------------------------------------------------------------
  if errors.any?
    puts "Bot startup failed — invalid config:"
    errors.each { |error| puts "  ✖ #{error}" }
    exit(1)
  end

  puts "Config validated successfully. Starting bot..."
end

# Registers signal handlers for graceful shutdown.
#
# These signals are common ways to stop a process:
#   - SIGTERM: usually sent by process managers or the stop command
#   - SIGINT: usually sent by Ctrl+C in the terminal
def setup_signal_handlers
  shutdown = proc do |signal_name|
    puts "Received #{signal_name}. Shutting down bot..."
    exit(0)
  end

  Signal.trap("TERM") { shutdown.call("SIGTERM") }
  Signal.trap("INT")  { shutdown.call("SIGINT") }
end

# Starts the bot runtime.
#
# Flow:
#   1. Validate config
#   2. Register signal handlers
#   3. Create a session ID for easier log tracing
#   4. Launch the infinite bot loop from spot.rb
def run_bot
  validate_config
  setup_signal_handlers

  session_id = "bot-#{(Time.now.to_f * 1000).to_i}"

  puts "*******************************************************"
  puts "LAUNCHING RUBY BOT"
  puts "*******************************************************"
  puts "Session ID: #{session_id}"

  initialize_bot(session_id)
end

run_bot if $PROGRAM_NAME == __FILE__

File: templates/python/runner.py
"""
===============================================================================
METAFIDE BOT RUNTIME — runner.py
===============================================================================

This file is the actual runtime launcher for the bot.

Responsibilities:
  1. Validate configuration before startup
  2. Register shutdown signal handlers
  3. Start the continuous bot loop from spot.py

This file is used in two modes:

  Foreground mode:
    python index.py run

  Background mode:
    python index.py start

Why this file exists:
  It separates "CLI process management" from "bot runtime logic".
  That keeps index.py focused on commands, while this file focuses on
  actually starting and running the bot safely.
===============================================================================
"""

from __future__ import annotations

import signal
import sys

from config import (
    ASSET,
    CURRENCY,
    ENABLE_EARLY_PRECISION,
    INTERVAL,
    MAX_ALLOWED_POSITIONS,
    METAFIDE_API_KEY,
    METAFIDE_USER_ADDRESS,
    NETWORK,
    POSITIONS_RANGE,
    PRICE_RANGES,
)
from spot import initialize


def validate_config() -> None:
    """
    Validates all config values before the bot starts.

    Why this matters:
      If config is invalid, we want to fail immediately with clear messages
      instead of letting the bot run in a broken or unsafe state.
    """
    errors: list[str] = []

    # -------------------------------------------------------------------------
    # Required fields
    # -------------------------------------------------------------------------
    if not METAFIDE_API_KEY:
        errors.append("METAFIDE_API_KEY is not set")

    if not METAFIDE_USER_ADDRESS:
        errors.append("METAFIDE_USER_ADDRESS is not set")

    # -------------------------------------------------------------------------
    # NETWORK validation
    # -------------------------------------------------------------------------
    valid_networks = ["mainnet", "testnet"]
    if NETWORK not in valid_networks:
        errors.append(
            f'NETWORK must be one of: {", ".join(valid_networks)} — got "{NETWORK}"'
        )

    # -------------------------------------------------------------------------
    # INTERVAL validation
    # -------------------------------------------------------------------------
    valid_intervals = [60, 3600, 23400, 86400]
    if INTERVAL not in valid_intervals:
        errors.append(
            f'INTERVAL must be one of: {", ".join(map(str, valid_intervals))} — got "{INTERVAL}"'
        )

    # -------------------------------------------------------------------------
    # Currency and asset validation
    # -------------------------------------------------------------------------
    if CURRENCY != "USDC":
        errors.append(f'CURRENCY must be "USDC" — got "{CURRENCY}"')

    if ASSET != "BTC_USDT":
        errors.append(f'ASSET must be "BTC_USDT" — got "{ASSET}"')

    # -------------------------------------------------------------------------
    # Position and feature flag validation
    # -------------------------------------------------------------------------
    if (
        not isinstance(MAX_ALLOWED_POSITIONS, int)
        or MAX_ALLOWED_POSITIONS < 1
        or MAX_ALLOWED_POSITIONS > 10
    ):
        errors.append(
            "MAX_ALLOWED_POSITIONS must be a number between 1 and 10 — "
            f'got "{MAX_ALLOWED_POSITIONS}"'
        )

    if not isinstance(ENABLE_EARLY_PRECISION, bool):
        errors.append(
            "ENABLE_EARLY_PRECISION must be true or false — "
            f'got "{ENABLE_EARLY_PRECISION}"'
        )

    # -------------------------------------------------------------------------
    # POSITIONS_RANGE validation
    # Each interval must have an allowed set of position sizes.
    # -------------------------------------------------------------------------
    position_minimums = {
        60: 0.1,
        3600: 1,
        23400: 5,
        86400: 5,
    }

    for interval in valid_intervals:
        range_values = POSITIONS_RANGE.get(interval)

        if not isinstance(range_values, list) or len(range_values) == 0:
            errors.append(f"POSITIONS_RANGE[{interval}] must be a non-empty array")
        else:
            minimum = position_minimums[interval]
            invalid = [
                value
                for value in range_values
                if not isinstance(value, (int, float)) or value < minimum
            ]

            if invalid:
                errors.append(
                    f"POSITIONS_RANGE[{interval}] values must be numbers >= "
                    f"{minimum} — invalid: [{', '.join(map(str, invalid))}]"
                )

    extra_position_keys = [
        key for key in POSITIONS_RANGE.keys() if key not in valid_intervals
    ]
    if extra_position_keys:
        errors.append(
            "POSITIONS_RANGE has unexpected keys: "
            f"[{', '.join(map(str, extra_position_keys))}] — only "
            f"{', '.join(map(str, valid_intervals))} are allowed"
        )

    # -------------------------------------------------------------------------
    # PRICE_RANGES validation
    # Each interval must define a numeric min/max randomization range.
    # -------------------------------------------------------------------------
    for interval in valid_intervals:
        price_range = PRICE_RANGES.get(interval)

        if not isinstance(price_range, dict):
            errors.append(f"PRICE_RANGES[{interval}] must be an object with min/max")
        else:
            if not isinstance(price_range.get("min"), (int, float)):
                errors.append(f"PRICE_RANGES[{interval}].min must be a number")

            if not isinstance(price_range.get("max"), (int, float)):
                errors.append(f"PRICE_RANGES[{interval}].max must be a number")

            if (
                isinstance(price_range.get("min"), (int, float))
                and isinstance(price_range.get("max"), (int, float))
                and price_range["min"] >= price_range["max"]
            ):
                errors.append(f"PRICE_RANGES[{interval}].min must be less than .max")

    extra_price_keys = [
        key for key in PRICE_RANGES.keys() if key not in valid_intervals
    ]
    if extra_price_keys:
        errors.append(
            "PRICE_RANGES has unexpected keys: "
            f"[{', '.join(map(str, extra_price_keys))}] — only "
            f"{', '.join(map(str, valid_intervals))} are allowed"
        )

    # -------------------------------------------------------------------------
    # Final result
    # -------------------------------------------------------------------------
    if errors:
        print("Bot startup failed — invalid config:")
        for error in errors:
            print(f"  ✖ {error}")
        sys.exit(1)

    print("Config validated successfully. Starting bot...")


def setup_signal_handlers() -> None:
    """
    Registers signal handlers for graceful shutdown.

    These signals are common ways to stop a process:
      - SIGTERM: usually sent by process managers or the stop command
      - SIGINT: usually sent by Ctrl+C in the terminal
    """

    def shutdown_handler(signum: int, _frame: object) -> None:
        signal_name = (
            "SIGTERM" if signum == signal.SIGTERM else
            "SIGINT" if signum == signal.SIGINT else
            f"signal {signum}"
        )
        print(f"Received {signal_name}. Shutting down bot...")
        sys.exit(0)

    signal.signal(signal.SIGTERM, shutdown_handler)
    signal.signal(signal.SIGINT, shutdown_handler)


def run_bot() -> None:
    """
    Starts the bot runtime.

    Flow:
      1. Validate config
      2. Register signal handlers
      3. Create a session ID for easier log tracing
      4. Launch the infinite bot loop from spot.py
    """
    import time

    validate_config()
    setup_signal_handlers()

    session_id = f"bot-{int(time.time() * 1000)}"

    print("*******************************************************")
    print("LAUNCHING PYTHON BOT")
    print("*******************************************************")
    print(f"Session ID: {session_id}")

    initialize(session_id)


if __name__ == "__main__":
    run_bot()